File: Surface/Surface/Puppeteer.m
//
//  Puppeteer.m
//  Surface
//
//  Created by Webchimp on 13/04/16.
//  Copyright © 2016 Webchimp. All rights reserved.
//

#import "Puppeteer.h"
#import "Helpers.h"

@implementation Puppeteer

- (id)init:(UIViewController *)controller
{
    self = [super init];
    if (self) {
        
        _context = controller;
        
        _pannelHeader = nil;
        _pannelFooter = nil;
        _pannelContent = nil;
        
        _pannelRoot = nil;
        
        _viewRoot = nil;
        _viewHeader = nil;
        _viewFooter = nil;
        _viewContent = nil;
        
        _footerBottons = [[NSMutableArray alloc] init];
        _footerPanels = [[NSMutableArray alloc] init];
        
        _panelMenuButton = nil;
        _viewMenuButton = nil;
    }
    return self;
}

- (SfPanel *)createLayout:(int)layout {
        
    _pannelRoot = [[SfPanel alloc] init];
    _pannelContent = [[SfPanel alloc] init];
    
    _viewRoot = [[UIView alloc] init];
    _viewContent = [[UIView alloc] init];
    
    [self.pannelRoot setView:self.viewRoot];
    [self.pannelContent setView:self.viewContent];
    
    [self.pannelRoot setKey:@"puppeteer-root"];
    [self.pannelContent setKey:@"puppeteer-content"];
    
    UINavigationController *nav = [[UINavigationController alloc] init];
    float headerHeight = nav.navigationBar.frame.size.height;
    
    switch (layout) {
        case PUPPET_HEADER_ONLY:
            
            _pannelHeader = [[SfPanel alloc] init];
            _viewHeader = [[UIView alloc] init];
            
            [self.pannelHeader setView:self.viewHeader];
            [self.pannelHeader setKey:@"puppeteer-header"];
            
            break;
            
        case PUPPET_FOOTER_ONLY:
            _pannelFooter = [[SfPanel alloc] init];
            _viewFooter = [[UIView alloc] init];
            
            [self.pannelFooter setView:self.viewFooter];
            [self.pannelFooter setKey:@"puppeteer-footer"];
            
            
            break;
            
        case PUPPET_HEADER_FOOTER:
            _pannelHeader = [[SfPanel alloc] init];
            _pannelFooter = [[SfPanel alloc] init];
            
            _viewHeader = [[UIView alloc] init];
            _viewFooter = [[UIView alloc] init];
            
            [self.pannelHeader setView:self.viewHeader];
            [self.pannelFooter setView:self.viewFooter];
            
            [self.pannelHeader setKey:@"puppeteer-header"];
            [self.pannelFooter setKey:@"puppeteer-footer"];
            
            break;
    }
    
    [self.context.view addSubview:self.viewRoot];
    if ( self.viewHeader != nil ) {
        [self.context.view addSubview:self.viewHeader];
        [self.pannelRoot append:self.pannelHeader];

    }
    if ( self.viewFooter != nil ) {
        [self.context.view addSubview:self.viewFooter];
        [self.pannelRoot append:self.pannelFooter];

    }
    
    [self.pannelRoot append:self.pannelContent];
    [self.context.view addSubview:self.viewContent];
    
    [self.pannelRoot setPosition:SF_POSITION_FIXED];
    [self.pannelRoot setSize:-100 andHeight:-100];
    [self.pannelRoot setOrigin:0 Right:SF_UNSET Bottom:SF_UNSET Left:0];
    //[self.pannelRoot setPadding:10 Right:0 Bottom:0 Left:0];
    
    switch (layout) {
        case PUPPET_HEADER_ONLY:
            
            [self.pannelHeader setPosition:SF_POSITION_FIXED];
            [self.pannelHeader setSize:-100 andHeight:headerHeight];
            [self.pannelHeader setOrigin:20 Right:SF_UNSET Bottom:SF_UNSET Left:0];
            [self.pannelHeader setPadding:5 Right:5 Bottom:5 Left:5];
            [self.pannelHeader setZIndex:10];
            
            [self.pannelContent setPosition:SF_POSITION_FIXED];
            [self.pannelContent setSize:-100 andHeight:0];
            [self.pannelContent setOrigin:headerHeight Right:SF_UNSET Bottom:0 Left:0];
            [self.pannelContent setPadding:5 Right:5 Bottom:5 Left:5];
            [self.pannelContent setZIndex:10];
            
        break;
        case PUPPET_FOOTER_ONLY:
        
            [self.pannelFooter setPosition:SF_POSITION_FIXED];
            [self.pannelFooter setSize:-100 andHeight:headerHeight];
            [self.pannelFooter setOrigin:SF_UNSET Right:SF_UNSET Bottom:0 Left:0];
            [self.pannelFooter setPadding:5 Right:5 Bottom:5 Left:5];
            [self.pannelFooter setZIndex:10];
            
            [self.pannelContent setPosition:SF_POSITION_FIXED];
            [self.pannelContent setSize:-100 andHeight:0];
            [self.pannelContent setOrigin:0 Right:SF_UNSET Bottom:headerHeight Left:0];
            [self.pannelContent setPadding:5 Right:5 Bottom:5 Left:5];
            [self.pannelContent setZIndex:10];
            
        break;
            
        case PUPPET_HEADER_FOOTER:
        
            [self.pannelHeader setPosition:SF_POSITION_FIXED];
            [self.pannelHeader setSize:-100 andHeight:headerHeight];
            [self.pannelHeader setOrigin:20 Right:SF_UNSET Bottom:SF_UNSET Left:0];
            [self.pannelHeader setPadding:5 Right:5 Bottom:5 Left:5];
            [self.pannelHeader setZIndex:10];
            
            [self.pannelFooter setPosition:SF_POSITION_FIXED];
            [self.pannelFooter setSize:-100 andHeight:headerHeight];
            [self.pannelFooter setOrigin:SF_UNSET Right:SF_UNSET Bottom:0 Left:0];
            [self.pannelFooter setPadding:5 Right:5 Bottom:5 Left:5];
            [self.pannelFooter setZIndex:10];
            
            [self.pannelContent setPosition:SF_POSITION_FIXED];
            [self.pannelContent setSize:-100 andHeight:0];
            [self.pannelContent setOrigin:headerHeight + 20 Right:SF_UNSET Bottom:headerHeight Left:0];
            [self.pannelContent setPadding:5 Right:5 Bottom:5 Left:5];
            [self.pannelContent setZIndex:10];
            
        break;
    }

    self.viewRoot.backgroundColor = [UIColor clearColor];
    self.viewHeader.backgroundColor = [Helpers rgbColor:76 :76 :76];
    self.viewFooter.backgroundColor = [Helpers rgbColor:76 :76 :76];
    self.viewContent.backgroundColor = [Helpers rgbColor:255 :255 :255];

    [self.pannelRoot update];
    
    return self.pannelContent;
}

- (SfPanel *)addFooterButton:(NSString *)caption :(UIImage *)image :(int)width :(SEL)listener {
    
    
    SfPanel *panel = nil;
    
    if ( self.pannelFooter != nil ) {
        
        panel = [[SfPanel alloc] init];
        UIButton *button = [UIButton buttonWithType:UIButtonTypeRoundedRect];
        
        if ( caption != nil || ![caption isEqualToString:@""] ) {
            [button setTitle:caption forState:UIControlStateNormal];
            button.backgroundColor = [UIColor whiteColor];
            [button setTintColor:[UIColor blueColor]];
        }
        
        if ( listener != nil ) {
            [button addTarget:self.context action:listener forControlEvents:UIControlEventTouchUpInside];
        }
        
        if ( image != nil ) {
            //
        }
        
        [panel setView:button];
        [panel setSize:width andHeight:self.pannelFooter.frame.height];
        [self.pannelFooter append:panel];
        [self.context.view addSubview:button];
        
        [self.footerBottons addObject:button];
        [self.footerPanels addObject:panel];
        
        [self.pannelFooter update];
        
    }
    
    return panel;
    
}


- (SfPanel *)addMenuButton:(NSString *)caption :(UIImage *)image :(int)width :(int)placement :(SEL)listener {
    
    if ( self.pannelHeader != nil ) {
        if ( self.panelMenuButton == nil ) {
            self.panelMenuButton = [[SfPanel alloc] init];
            self.viewMenuButton = [UIButton buttonWithType:UIButtonTypeRoundedRect];
            
            if ( caption != nil || ![caption isEqualToString:@""] ) {
                
                [self.viewMenuButton setTitle:caption forState:UIControlStateNormal];
                self.viewMenuButton.backgroundColor = [UIColor whiteColor];
                [self.viewMenuButton setTintColor:[UIColor blueColor]];
            }
            
            if ( listener != nil ) {
                [self.viewMenuButton addTarget:self.context action:listener forControlEvents:UIControlEventTouchUpInside];
            }
            
            [self.panelMenuButton setView:self.viewMenuButton];
            [self.panelMenuButton setSize:width andHeight:self.pannelHeader.frame.height];
            [self.pannelHeader append:self.panelMenuButton];
            [self.context.view addSubview:self.viewMenuButton];
            
            [self.panelMenuButton setPosition:SF_POSITION_ABSOLUTE];
            
            if ( placement == PUPPET_MENU_LEFT ) {
                [self.panelMenuButton setOrigin:0 Right:SF_UNSET Bottom:SF_UNSET Left:0];
            } else {
                [self.panelMenuButton setOrigin:0 Right:0 Bottom:SF_UNSET Left:SF_UNSET];
            }
            
            [self.pannelHeader update];
            
        }
    }
    
    return self.panelMenuButton;
}


@end

File: Java/Puppeteer/Puppeteer.java
package com.appbuilders.quienesestepoke.libraries.Surface.Puppeteer;

import android.app.Activity;
import android.content.Context;
import android.graphics.Color;
import android.util.TypedValue;
import android.view.View;
import android.widget.AbsoluteLayout;
import android.widget.Button;

import com.appbuilders.quienesestepoke.libraries.Surface.SfPanel;

import java.util.ArrayList;

public class Puppeteer {

    AbsoluteLayout baseLayout;
    Context context;
    SfPanel panelRoot;
    SfPanel panelHeader;
    SfPanel panelFooter;
    SfPanel panelContent;
    View viewRoot;
    View viewHeader;
    View viewFooter;
    View viewContent;
    ArrayList<View> footerButtons;
    ArrayList<SfPanel> footerPanels;
    SfPanel panelMenuButton;
    Button viewMenuButton;

    public static final int PUPPET_HEADER_ONLY   = 0;
    public static final int PUPPET_FOOTER_ONLY   = 1;
    public static final int PUPPET_HEADER_FOOTER = 2;

    public static final int PUPPET_MENU_LEFT  = 0;
    public static final int PUPPET_MENU_RIGHT = 1;

    public Puppeteer(Context context) {
        this.context= context;
        this.baseLayout = null;
        this.panelRoot = null;
        this.panelHeader = null;
        this.panelFooter = null;
        this.panelContent = null;
        this.viewRoot = null;
        this.viewHeader = null;
        this.viewFooter = null;
        this.viewContent = null;
        this.footerButtons = new ArrayList<View>();
        this.footerPanels = new ArrayList<SfPanel>();
        this.panelMenuButton = null;
        this.viewMenuButton = null;
    }

    public SfPanel addMenuButton(String caption, int image, int width, int placement, View.OnClickListener listener) {
        if (this.panelHeader != null) {
            if (this.panelMenuButton == null) {
                panelMenuButton = new SfPanel();
                viewMenuButton = new Button(this.context);
            }
            // Button stuff
            if (caption != null && caption.compareTo("") != 0) {
                viewMenuButton.setText(caption);
            }
            if (image > 0) {
                viewMenuButton.setBackgroundResource(image);
            }
            if (listener != null) {
                viewMenuButton.setOnClickListener(listener);
            }
            // Panel stuff
            panelMenuButton.setView(viewMenuButton);
            panelMenuButton.setSize(width, this.panelHeader.frame.height);
            this.panelHeader.append(panelMenuButton);
            this.baseLayout.addView(viewMenuButton);
            //
            panelMenuButton.setPosition(SfPanel.SF_POSITION_ABSOLUTE);
            if (placement == PUPPET_MENU_LEFT) {
                panelMenuButton.setOrigin(0, SfPanel.SF_UNSET, SfPanel.SF_UNSET, 0);
            } else {
                panelMenuButton.setOrigin(0, 0, SfPanel.SF_UNSET, SfPanel.SF_UNSET);
            }
            // Update
            this.panelHeader.update(this.context);
        }
        return panelMenuButton;
    }

    public SfPanel addFooterButton(String caption, int image, int width, View.OnClickListener listener) {
        SfPanel panel = null;
        Button button = null;
        if (this.panelFooter != null) {
            panel = new SfPanel();
            button = new Button(this.context);
            // Button stuff
            if (caption != null && caption.compareTo("") != 0) {
                button.setText(caption);
            }
            if (image > 0) {
                button.setBackgroundResource(image);
            }
            if (listener != null) {
                button.setOnClickListener(listener);
            }
            // Panel stuff
            panel.setView(button);
            panel.setSize(width, this.panelFooter.frame.height);
            this.panelFooter.append(panel);
            this.baseLayout.addView(button);
            // Add them to the arrays
            this.footerPanels.add(panel);
            this.footerButtons.add(button);
            // Update
            this.panelFooter.update(this.context);
        }
        return panel;
    }

    public SfPanel getFooterButton() {
        return this.panelFooter;
    }

    public SfPanel removeFooterButton() {
        return this.panelFooter;
    }

    public SfPanel clearFooterButtons() {
        return this.panelFooter;
    }

    public SfPanel createLayout(int layout) {
        TypedValue tv = new TypedValue();
        int headerHeight = 0;
        this.context.getTheme().resolveAttribute(android.R.attr.actionBarSize, tv, true);
        headerHeight = context.getResources().getDimensionPixelSize(tv.resourceId) + 12;
        // Create the base layout
        this.baseLayout = new AbsoluteLayout(this.context);
        // Create root and content panels/views
        this.panelRoot = new SfPanel();
        this.panelContent = new SfPanel();
        this.viewRoot = new View(this.context);
        this.viewContent = new View(this.context);
        // Set panel views
        this.panelRoot.setView(viewRoot);
        this.panelContent.setView(viewContent);
        this.panelRoot.setKey("puppeteer-root");
        this.panelContent.setKey("puppeteer-content");
        // Create header and/or footer panels/views
        switch (layout) {
            case PUPPET_HEADER_ONLY:
                this.panelHeader = new SfPanel();
                this.viewHeader = new View(this.context);
                // Set panel view
                this.panelHeader.setView(viewHeader);
                this.panelHeader.setKey("puppeteer-header");
                break;
            case PUPPET_FOOTER_ONLY:
                this.panelFooter = new SfPanel();
                this.viewFooter = new View(this.context);
                // Set panel view
                this.panelFooter.setView(viewFooter);
                this.panelFooter.setKey("puppeteer-footer");
                break;
            case PUPPET_HEADER_FOOTER:
                this.panelHeader = new SfPanel();
                this.panelFooter = new SfPanel();
                this.viewHeader = new View(this.context);
                this.viewFooter = new View(this.context);
                // Set panel views
                this.panelHeader.setView(viewHeader);
                this.panelFooter.setView(viewFooter);
                this.panelHeader.setKey("puppeteer-header");
                this.panelFooter.setKey("puppeteer-footer");
                break;
        }
        // Add views into base layout and create panel hierarchy
        this.baseLayout.addView(viewRoot);
        if (this.viewHeader != null) {
            this.baseLayout.addView(viewHeader);
            this.panelRoot.append(panelHeader);
        }
        if (this.viewFooter != null) {
            this.baseLayout.addView(viewFooter);
            this.panelRoot.append(panelFooter);
        }
        this.baseLayout.addView(viewContent);
        this.panelRoot.append(panelContent);
        // Size and position panels
        this.panelRoot.setPosition(SfPanel.SF_POSITION_FIXED)
                .setSize(-100, -100)
                .setOrigin(0, SfPanel.SF_UNSET, SfPanel.SF_UNSET, 0);
        switch (layout) {
            case PUPPET_HEADER_ONLY:
                this.panelHeader.setPosition(SfPanel.SF_POSITION_FIXED)
                        .setSize(-100, headerHeight)
                        .setOrigin(0, SfPanel.SF_UNSET, SfPanel.SF_UNSET, 0)
                        .setPadding(5, 5, 5, 5)
                        .setZIndex(10);
                this.panelContent.setPosition(SfPanel.SF_POSITION_FIXED)
                        .setSize(-100, 0)
                        .setOrigin(headerHeight, SfPanel.SF_UNSET, 0, 0)
                        .setPadding(5, 5, 5, 5)
                        .setZIndex(10);
                break;
            case PUPPET_FOOTER_ONLY:
                this.panelFooter.setPosition(SfPanel.SF_POSITION_FIXED)
                        .setSize(-100, headerHeight)
                        .setOrigin(SfPanel.SF_UNSET, SfPanel.SF_UNSET, 0, 0)
                        .setPadding(5, 5, 5, 5)
                        .setZIndex(10);
                this.panelContent.setPosition(SfPanel.SF_POSITION_FIXED)
                        .setSize(-100, 0)
                        .setOrigin(0, SfPanel.SF_UNSET, headerHeight, 0)
                        .setPadding(5, 5, 5, 5)
                        .setZIndex(10);
                break;
            case PUPPET_HEADER_FOOTER:
                this.panelHeader.setPosition(SfPanel.SF_POSITION_FIXED)
                        .setSize(-100, headerHeight)
                        .setOrigin(0, SfPanel.SF_UNSET, SfPanel.SF_UNSET, 0)
                        .setPadding(5, 5, 5, 5)
                        .setZIndex(10);
                this.panelFooter.setPosition(SfPanel.SF_POSITION_FIXED)
                        .setSize(-100, headerHeight)
                        .setOrigin(SfPanel.SF_UNSET, SfPanel.SF_UNSET, 0, 0)
                        .setPadding(5, 5, 5, 5)
                        .setZIndex(10);
                this.panelContent.setPosition(SfPanel.SF_POSITION_FIXED)
                        .setSize(-100, 0)
                        .setOrigin(headerHeight, SfPanel.SF_UNSET, headerHeight, 0)
                        .setPadding(5, 5, 5, 5)
                        .setZIndex(10);
                break;
        }
        //
        Activity activity = (Activity) this.context;
        activity.setContentView(this.baseLayout);
        // For debug purposes only
        this.viewRoot.setBackgroundColor(Color.rgb(220, 220, 220));
        if (this.viewHeader != null) {
            this.viewHeader.setBackgroundColor(Color.rgb(76, 76, 76));
        }
        if (this.viewFooter != null) {
            this.viewFooter.setBackgroundColor(Color.rgb(76, 76, 76));
        }
        this.viewContent.setBackgroundColor(Color.rgb(255, 255, 255));
        //
        this.panelRoot.update(this.context);
        //
        return this.panelContent;
    }
}